### DISCUSSION TITLE
password update error: Password update requires reauthentication.

### DISCUSSION THREAD

**Discussion by ngsmcphrsn** [supabase/supabase#7310] [Questions] [answered]
Very new to supabase, getting a password update error in a simple sveltekit setup.

Trying to enable password update/creation for users signed in and authenticated via magic link. The `const { data, error } = await supabase.auth.update({ password: "password" });` workflow throws a `401 (Unauthorized) Password update requires reauthentication` error.

```
const updatePassword = async () => {
  try {
    const { user, error } = await supabase.auth.update({
      password: "password"
    });
    if (error) throw error;
    updateMessage = `Password updated: ${new Date()}`;
  } catch (error: any) {
    updateMessage = `${error.error_description || error.message}`;
  } finally {
    loading = false;
  }
};
```

The same approach is successful for updating email (triggers the 2 step confirmation path). 

Additional wrinkle. Doesn't return the error when passed ` password: undefined`. And it successfully updates metadata when the payload is: ```
{
  password: undefined,
  data: { hello: 'world' }
} ```

**Comment by GaryAustin1** [comment]
Is the user logged in?
If you need to reset a forgotten password you use: https://supabase.com/docs/reference/javascript/auth-api-resetpasswordforemail

**Reply by ngsmcphrsn** [reply]
Yep, user is logged in, though only with the magic link (`subabase.auth.user()` returns an object that with `aud: "authenticated"`). The case I'm looking at is an invited user with no password set, so the password reset flow feels wrong. 

`supabase.auth.update()` works as expected when initiating the change email workflow and adding metadata. It only returns an error when passed `password`.

fwiw, I've also tried the `.updateUser()` method, like so:

```
const accessToken = supabase.auth.session()?.access_token;
try {
  const { user, error } = await supabase.auth.api.updateUser(accessToken, {
    password: 'password'
});
```

It returns the same error.

**Comment by GaryAustin1** [comment]
If you get a chance you might try a signup with email/password and then your same password code.  Wonder if the magic link only is an issue.

**Reply by ngsmcphrsn** [reply]
Ah. Right. Just tried that and got the same error. So not just a magic link thing.

Am I missing something super obvious...?

**Reply by GaryAustin1** [reply]
Sure is acting like you are not signed in, or at least the jwt is not set when you make the call.

**Reply by GaryAustin1** [reply]
I just tested it and it works:
```
const {session,user,error} = await supabase.auth.signIn({
            email: 'me@me.com',
            password:'test1234'
        })
       console.log('login',session,user,error)
        const result = await supabase.auth.update({
            password:'test1245'
        })
        console.log('change',result)
```

**Reply by ngsmcphrsn** [reply]
Copy. Thanks for doing that. I must be doing something janky with the token, though it seems to be setting correctly between signIn() and signOut().

Going to close this, since it looks like my error. Thanks for the help.

**Reply by ngsmcphrsn** [reply]
Ok. Still getting the error. I've stripped this all the way down to be as minimal as possible. Spun up a fresh SvelteKit project containing nothing but the supabaseClient, the .env, a bit of markup, and these functions:
```
<script>
  import { supabase } from '$lib/supabaseClient';

  let email = '';
  let password = '';

  const handleSignUp = async () => {
    const { user, session, error } = await supabase.auth.signUp({
      email,
      password
    });
  };

  const updatePassword = async () => {
    const { user, error } = await supabase.auth.update({
      password
    });
  };

  const handleSignIn = async () => {
    const { user, session, error } = await supabase.auth.signIn({
      email,
      password
    });
  };
</script>
```

(*though using the existing supabase db, which was initialized with "User Management Starter", per the docs)

Password update is still throwing exactly the same error.:
```
"code": 401,
"msg": "Password update requires reauthentication."
```

I'm on:
`"@supabase/supabase-js": "^1.35.3"`
`"@sveltejs/kit": "next"`
`"svelte": "^3.44.0"`

**Reply by GaryAustin1** [reply]
I don't know if a corrupted auth.users table could cause the error or not.  If you hard code a call to your signIn function and then update function that does that work?  I'm on 1.35.3.  I use sveltekit, but not for the my quick test environment.  The session data should be stored in the supabase object from signin to update so should not depend on local storage, etc.

**Reply by ngsmcphrsn** [reply]
Ok. Thanks again. Hard coding produced the same error. So I finally initialized a new database instance, and..it seems to work just fine. No issues in any of the original code. It must be a corrupted table/db.

Not sure I love the error it threw though.

**Reply by GaryAustin1** [reply]
Sigh.  Really should be tracked down by SB if the user table can be damaged.  IF this was for more than one user, including new users, in the table that would be really bad.  If it was for one user, then maybe some early signUp method is incompatible once it happens.

**Comment by ngsmcphrsn** [comment, marked as answer]
So. After all that, it's an auth config setting that I somehow missed.

Still, not sure what the canonical reauthentication workflow is.

**Reply by Darren120** [reply]
this is not an answer, you turned off a feature and called it a solution.

**Comment by BlockInside** [comment]
I'm getting same error too. Can you a little explain more what was the problem @ngsmcphrsn

**Reply by BlockInside** [reply]
I saw a parameter in gotrue; SECURITY_UPDATE_PASSWORD_REQUIRE_REAUTHENTICATION
and studio has this parameter; https://github.com/supabase/supabase/commit/c434561caba54bd4170cb0743a163e0f0d2c19af#diff-a47903ffa1ce3edb0bec27ac6c4c9877e074532bef408e8f8fca62b21ffe6aa4R148

```go
SECURITY_UPDATE_PASSWORD_REQUIRE_REAUTHENTICATION: model.SECURITY_UPDATE_PASSWORD_REQUIRE_REAUTHENTICATION || false,
```

But this "require reauthenticate" toggle is not shown in supabase dashboard or am i missing?

**Reply by ngsmcphrsn** [reply]
I think that's right. I'd missed the "allow password update" toggle in an old version of the supabase UI. Toggling it fixed the error, and the update works as expected.

I can't find the toggle in the new UI though.

**Reply by BlockInside** [reply]
That means, your previos work is working as well @ngsmcphrsn 

Oh my... There is nothing we can do right now then.
1- I dont know how to reauthenticate a user. Documentation is lacking about this.
2- Dashboard UI dont have a "require reauthentication" toggle switch.

**Reply by jishnair** [reply]
I am also getting the same error, after authenticating the user and calling updateUser function.
```
const { error, data } = await supabase.auth.api
      .updateUser(access_token, { password : new_password })
```
Response: 
```
{
  "code": 401,
  "msg": "Password update requires reauthentication."
}
```

**Reply by jojomatik** [reply]
Have you guys since found a solution? Could someone from @supabase confirm that this feature is missing?

Would it be possible to open this discussion again?

**Reply by jojomatik** [reply]
> Have you guys since found a solution? Could someone from @supabase confirm that this feature is missing?
> 
> Would it be possible to open this discussion again?

It seems like the problem does not appear in a newly created project. I've created a support ticket to try and fix my old project.

**Reply by BlockInside** [reply]
@jojomatik
I will try again today.
Creating a new project could be a solution, yes. But it couldn't be so hard to integrate "new feature"/"bug fix" to old projects.

**Reply by jojomatik** [reply]
I've been in contact with support and they have not been able to fix the old project. They provided me with a [migration script](https://github.com/mansueli/Supa-Migrate) that dumps the entire old project into a file and uploads it to the new project, but it didn't really work for me. 

My current plan is to create a fresh project and migrate all of our manually created tables to that project.

**Comment by versoworks** [comment]
@GaryAustin1 I ran in to this error today. I had enabled the `Secure password change` option in Supabase Studio and so was expecting to re-authenticate the user. I can't seem to find the method to do so in the supabase-js client. Is it not exposed yet? I see in the [docs ](https://github.com/supabase/gotrue#get-reauthenticate) that re-authentication is available already in the api. I've temporarily disabled the `Secure password change` option to work around the error but would prefer to have it enabled.

**Reply by GaryAustin1** [reply]
The setting just means you need to login then you can change the password.  You can't have an old session being refreshed.

**Reply by versoworks** [reply]
@GaryAustin1 I just tried that and the error persists.

**Reply by GaryAustin1** [reply]
Not sure then.  Sounds like you need to get help debugging your code, or if you are sure it is a Supabase error, post an issue.
Tagging on to this older post does not help you get exposure except to users on it.   If you start a new discussion or generate an issue you should post more about your environment and code.   Unless something broke very very recently, resetting password userUpdate() right after a new signin (not a refresh of session) should work.

**Reply by t-lock** [reply]
> Unless something broke very very recently, resetting password userUpdate() right after a new signin (not a refresh of session) should work.

This appears to be the case. I have a brand new project and get this error. Then again I can't even locate a "Secure password change" option in Studio...

**Reply by versoworks** [reply]
> > Unless something broke very very recently, resetting password userUpdate() right after a new signin (not a refresh of session) should work.
> 
> This appears to be the case. I have a brand new project and get this error. Then again I can't even locate a "Secure password change" option in Studio...

It's not the case; the advice on here is wrong. I opened an [issue](https://github.com/supabase/gotrue-js/issues/680) and it has been [resolved](https://github.com/supabase/gotrue-js/pull/688).
